File: implementation/scipamato/core/sync/src/main/java/ch/difty/scipamato/core/sync/jobs/language/LanguageItemWriter.java
package ch.difty.scipamato.core.sync.jobs.language;

import static ch.difty.scipamato.publ.db.public_.tables.Language.LANGUAGE;

import org.jooq.DSLContext;
import org.springframework.batch.item.ItemWriter;

import ch.difty.scipamato.core.sync.jobs.ScipamatoItemWriter;

/**
 * {@link ItemWriter} implementation to synchronize languages from scipamato-core
 * to public.
 *
 * <ul>
 * <li>Takes care of inserts and updates.</li>
 * <li>Every record will be updated at least with the current timestamp in
 * last_synched.</li>
 * </ul>
 *
 * @author u.joss
 */
public class LanguageItemWriter extends ScipamatoItemWriter<PublicLanguage> {

    @SuppressWarnings("WeakerAccess")
    public LanguageItemWriter(final DSLContext jooqDslContextPublic) {
        super(jooqDslContextPublic, "language");
    }

    @Override
    protected int executeUpdate(final PublicLanguage l) {
        return getDslContext()
            .insertInto(LANGUAGE)
            .columns(LANGUAGE.CODE, LANGUAGE.MAIN_LANGUAGE, LANGUAGE.LAST_SYNCHED)
            .values(l.getCode(), l.getMainLanguage(), l.getLastSynched())
            .onConflict(LANGUAGE.CODE)
            .doUpdate()
            .set(LANGUAGE.CODE, l.getCode())
            .set(LANGUAGE.MAIN_LANGUAGE, l.getMainLanguage())
            .set(LANGUAGE.LAST_SYNCHED, l.getLastSynched())
            .where(LANGUAGE.CODE.eq(l.getCode()))
            .execute();
    }

}

File: core/core-sync/src/main/kotlin/ch/difty/scipamato/core/sync/jobs/language/LanguageItemWriter.kt
package ch.difty.scipamato.core.sync.jobs.language

import ch.difty.scipamato.core.sync.PublicLanguage
import org.jooq.DSLContext
import ch.difty.scipamato.core.sync.jobs.ScipamatoItemWriter
import ch.difty.scipamato.publ.db.tables.Language

/**
 * [ScipamatoItemWriter] implementation to synchronize languages from scipamato-core to public.
 *
 *  * Takes care of inserts and updates.
 *  * Every record will be updated at least with the current timestamp in last_synched.
 */
class LanguageItemWriter(jooqDslContextPublic: DSLContext) : ScipamatoItemWriter<PublicLanguage>(jooqDslContextPublic, "language") {
    override fun executeUpdate(i: PublicLanguage): Int =
        dslContext
            .insertInto(Language.LANGUAGE)
            .columns(Language.LANGUAGE.CODE, Language.LANGUAGE.MAIN_LANGUAGE, Language.LANGUAGE.LAST_SYNCHED)
            .values(i.code, i.mainLanguage, i.lastSynched)
            .onConflict(Language.LANGUAGE.CODE)
            .doUpdate()
            .set(Language.LANGUAGE.CODE, i.code)
            .set(Language.LANGUAGE.MAIN_LANGUAGE, i.mainLanguage)
            .set(Language.LANGUAGE.LAST_SYNCHED, i.lastSynched)
            .where(Language.LANGUAGE.CODE.eq(i.code))
            .execute()
}
